### PR TITLE
Add support for custom backends

### PR DESCRIPTION
**Pull request by hishnash**
We use a `Capsule` to pass the `git_repository` struct through python. 

This `Capsule` does not need a destructor since libgit will free the `git_repository` struct when it needs to.

this addresses some of the needs for https://github.com/libgit2/pygit2/issues/76

### CODE REVIEW

File: pygit2/repository.py [hunk 1]
```diff
@@ -915,3 +904,22 @@ def set_ident(self, name, email):
...
+
+class Repository(BaseRepository):
+    def __init__(self, path, *args, **kwargs):
+        if not isinstance(path, six.string_types):
+            path = path.decode('utf-8')
+
+        path_backend = init_file_backend(path)
+        super(Repository, self).__init__(backend=path_backend, *args, **kwargs)
+        self._common_init()
+
```
Comment (jdavid) [inline pygit2/repository.py]: ``_common_init`` would be called twice, first from the super class, and then here

File: src/pygit2.c [hunk 2]
```diff
@@ -144,7 +144,43 @@ hash(PyObject *self, PyObject *args)
...
+PyObject *
+init_file_backend(PyObject *self, PyObject *args)
+{
+    const char* path;
+	int err = GIT_OK;
+	git_repository *repository = NULL;
+	if (!PyArg_ParseTuple(args, "s", &path)) {
+		return NULL;
+	};
+
```
Comment (jdavid) [inline src/pygit2.c]: could you fix the indentation please? here and elsewhere in the C code

File: pygit2/repository.py (hunk 1 above)
Comment (hishnash) [inline pygit2/repository.py]: fixed

File: src/pygit2.c (hunk 2 above)
Comment (hishnash) [inline src/pygit2.c]: fixed

**Comment by webknjaz** [comment]
Hey @hishnash, is it possible to use this with non-file backends? like mysql or so

**Comment by hishnash** [comment]
@webknjaz yes this was done to allow for https://github.com/luckymarmot/GitPagedArchiveBackend 

here we are just using a custom backed for the ODB portion but you could use something very simlare to get a RefDB backend loaded as well.

**Comment by webknjaz** [comment]
Thank you for the reference, I'll try that :)

**Pull request merged**